_posts/2018-02-15-continuous-deployment-concourse.md

```diff
@@ -0,0 +1,286 @@
+
+Another massive argument is the scale at which Concourse can operate. As long as you have resources to schedule your containers and volumes, the system will scale. Where I work, we have Concourse workers on an autoscaling group in Google Cloud. At peak traffic we can have dozens of concurrent builds, spinning hundreds of containers and sometimes thousands of volumes.
+
+## Concourse CI notions
+
+- *Resources* are external assets, a Docker image on a registry, a github repository, a slack hook, you name it. They will almost all have checks to trigger jobs, inputs to retrieve and outputs to update/push the resource.
+- *Jobs* contain tasks, these tasks will be executed together in parallel or in sequence on the same node. You will be able to pass inputs and outputs from one task to another.
+- *Tasks* are like functions, with inputs and outputs. Since it's a container you can choose which image it will use to fit your needs.
+
+The *pipeline* contains all the jobs necessary to fulfill a purpose. When a resource gets updated it will trigger a *build* which is the logical execution of the job's tasks.
+
+## Build images
+
+### Docker image resource definition:
+```
+- name: foobar-app-image
+  type: docker-image
+  source:
+    repository: gcr.io/foobar/foobar_app
+    username: _json_key
+    password: {{gcloud_service_key}}
+```
+
+### Build and push the image:
+```
+- put: foobar-app-image
+  params:
+    build: artifact-app/
+    cache: true
+    cache_tag: latest
+    tag: metadata/short-commit-hash
+    tag_as_latest: true
+```
+Note: 
+- the artifact-app/ volume contains a Dockerfile and dependencies to build the image 
+- the `latest` tag will be cached on disk, if we have a lot of builds this can save time!
+- The metadata is a simple task that uses a `git` Docker image to retrieve the git short commit hash:
+    ```
+    ---
+    platform: linux
+
+    image_resource:
+      type: docker-image
+      source: {repository: greenwall/git}
+
+    params:
+      TERM: xterm-256color
+
+    inputs:
+      - name: code-base
+
+    outputs:
+      - name: metadata
+
+    run:
+      path: /bin/sh
+      args:
+      - -exc
+      - |
+        GIT_SHORT_HASH=`cd code-base && git log --pretty=format:'%h' -n 1`
+        echo $GIT_SHORT_HASH > metadata/short-commit-hash
+    ```
+
+## Run unit tests on pre-built container
+We could easily reuse that image for unit testing, tests that should not depend on anything else than the code itself. Here's an example on how to do it with php-unit:
```

Comment:
- `tests that should not depend` => `tests should not depend`
- _[STYLE]_ php-unit could be wrapped in backticks as this is the library name.